File: app/models/competitor.py
"""Competitor and public-link parsing task models for Phase 3A."""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import TYPE_CHECKING, Any

from sqlalchemy import JSON, DateTime, Enum as SqlEnum, ForeignKey, Integer, Numeric, String, Text, func, text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base
from app.models.store import Platform

if TYPE_CHECKING:
    from app.models.product import Product


class PublicLinkParseTaskStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"


class Competitor(Base):
    __tablename__ = "competitors"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    product_id: Mapped[int] = mapped_column(
        ForeignKey("products.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    platform: Mapped[Platform] = mapped_column(
        SqlEnum(
            Platform,
            name="competitor_platform_type",
            native_enum=False,
            create_constraint=True,
            validate_strings=True,
            values_callable=lambda enum_class: [item.value for item in enum_class],
        ),
        nullable=False,
        index=True,
    )
    url: Mapped[str] = mapped_column(String(2048), nullable=False)
    price: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    sales_hint: Mapped[str | None] = mapped_column(String(200))
    title: Mapped[str | None] = mapped_column(String(300))
    main_image: Mapped[str | None] = mapped_column(String(2048))
    selling_points: Mapped[list[str]] = mapped_column(JSON, default=list, nullable=False)
    review_keywords: Mapped[list[str]] = mapped_column(JSON, default=list, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    product: Mapped["Product"] = relationship(back_populates="competitors")


class PublicLinkParseTask(Base):
    __tablename__ = "public_link_parse_tasks"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    product_id: Mapped[int] = mapped_column(
        ForeignKey("products.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    source_url: Mapped[str] = mapped_column(String(2048), nullable=False)
    task_status: Mapped[PublicLinkParseTaskStatus] = mapped_column(
        SqlEnum(
            PublicLinkParseTaskStatus,
            name="public_link_parse_task_status",
            native_enum=False,
            create_constraint=True,
            validate_strings=True,
            values_callable=lambda enum_class: [item.value for item in enum_class],
        ),
        default=PublicLinkParseTaskStatus.PENDING,
        server_default=text("'pending'"),
        nullable=False,
        index=True,
    )
    attempts: Mapped[int] = mapped_column(Integer, default=0, server_default=text("0"), nullable=False)
    result_json: Mapped[dict[str, Any] | None] = mapped_column(JSON)
    error_message: Mapped[str | None] = mapped_column(Text)
    confirmed_competitor_id: Mapped[int | None] = mapped_column(
        ForeignKey("competitors.id", ondelete="RESTRICT"),
        unique=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    product: Mapped["Product"] = relationship(back_populates="link_parse_tasks")

File: frontend/src/types/competitor.ts
import type { Platform } from './store'

export interface Competitor {
  id: number
  product_id: number
  name: string
  platform: Platform
  url: string
  price: string | null
  sales_hint: string | null
  title: string | null
  main_image: string | null
  selling_points: string[]
  review_keywords: string[]
  created_at: string
  updated_at: string
}

export interface CompetitorCreate {
  name: string
  platform: Platform
  url: string
  price?: string | null
  sales_hint?: string | null
  title?: string | null
  main_image?: string | null
  selling_points?: string[]
  review_keywords?: string[]
}

export type CompetitorUpdate = Partial<CompetitorCreate>

export type PublicLinkParseTaskStatus = 'pending' | 'running' | 'succeeded' | 'failed'

export interface PublicLinkCandidate {
  name?: string
  platform?: Platform
  price?: string | null
  sales_hint?: string | null
  title?: string | null
  main_image?: string | null
  selling_points?: string[]
  review_keywords?: string[]
  data_source?: string
}

export interface PublicLinkParseTask {
  id: number
  product_id: number
  source_url: string
  task_status: PublicLinkParseTaskStatus
  attempts: number
  result_json: PublicLinkCandidate | null
  error_message: string | null
  confirmed_competitor_id: number | null
  created_at: string
  updated_at: string
}
